[DAG] Fold nested add(add(reduce(a), b), add(reduce(c), d))

davemgreen opened the pull request:
This patch reassociates `add(add(vecreduce(a), b), add(vecreduce(c), d))` into `add(vecreduce(add(a, c)), add(b, d))`, to combine the reductions into a single node. This comes up after unrolling vectorized loops.

There is another small change to move reassociateReduction inside fadd outside of a AllowNewConst block, as new constants will not be created and it should be OK to perform the combine later after legalization.

RKSimon commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -1329,6 +1329,38 @@ SDValue DAGCombiner::reassociateReduction(unsigned RedOpc, unsigned Opc,
+    if (N00.getOpcode() != RedOpc && N01.getOpcode() == RedOpc)
+      std::swap(N00, N01);
+    if (N00.getOpcode() == RedOpc && N01.getOpcode() != RedOpc &&
+        N00->hasOneUse()) {
```
I'm pretty sure you can use sd_match to handle all of this matching (including the operand swapping).

davemgreen replied:
Oh that does sound like a better approach. Thanks.

david-arm commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -17107,12 +17129,15 @@ SDValue DAGCombiner::visitFADD(SDNode *N) {
+  } // enable-unsafe-fp-math && AllowNewConst
 
+  if (((Options.UnsafeFPMath && Options.NoSignedZerosFPMath) ||
+       (Flags.hasAllowReassociation() && Flags.hasNoSignedZeros()))) {
```
Does removing the `AllowNewConst` restriction from the reassociateReduction call have any effect on the tests. Just wondering if this is worth a separate PR?

david-arm commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -1329,6 +1329,28 @@ SDValue DAGCombiner::reassociateReduction(unsigned RedOpc, unsigned Opc,
+                                  m_Value(D)))) &&
+      !sd_match(B, m_UnaryOp(RedOpc, m_Value())) &&
+      !sd_match(D, m_UnaryOp(RedOpc, m_Value())) &&
+      A.getValueType() == C.getValueType() &&
```
Is it even possible for these to be different types? I assume this would have to be some sort of extending reduction where the amount of extension was different.

david-arm commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -1329,6 +1329,28 @@ SDValue DAGCombiner::reassociateReduction(unsigned RedOpc, unsigned Opc,
+  // single node.
+  SDValue A, B, C, D;
+  if (sd_match(N0,
+               m_OneUse(m_c_BinOp(Opc, m_OneUse(m_UnaryOp(RedOpc, m_Value(A))),
```
Do we need to worry about the reductions having different flags on the operations? Are they expected to match the flags passed in?

david-arm replied:
Thanks for this! Is it worth adding an extra test in CodeGen/AArch64/double_reduct.ll, i.e. something like `nested_mul_f32_mixed_reassoc`?

david-arm reviewed (commented):
The codegen changes look good! Do you have any examples where this type of pattern comes up? I had a few other minor comments.

david-arm commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -1329,6 +1329,38 @@ SDValue DAGCombiner::reassociateReduction(unsigned RedOpc, unsigned Opc,
+      return SDValue();
+    SelectionDAG::FlagInserter FlagsInserter(DAG, Flags);
+    SDValue Op = DAG.getNode(Opc, DL, A.getValueType(), A, C);
+    SDValue Red = DAG.getNode(RedOpc, DL, VT, Op);
```
Sorry to be picky, but is there still a problem here with flags for FP operations? Perhaps I'm just being overly cautious, but I'm thinking of an example where the flags on the top level `fadd` may be different to those on the two individual reductions, etc. So effectively we could end up promoting the reduction to have flags that it didn't have previously. I must admit I'm a bit unsure about what happens in general for DAG combines in cases like this:

```
  %r1 = call fast float @llvm.vector.reduce.fadd.f32.v4f32(float -0.0, <4 x float> %a)
  %a1 = fadd fast float %r1, %c
  %r2 = call reassoc float @llvm.vector.reduce.fadd.f32.v4f32(float -0.0, <4 x float> %b)
  %a2 = fadd reassoc float %r2, %d
  %r = fadd fast float %a1, %a2
```

Might it be simpler to require all nested operations to have the same flags as the `Flags` passed in if that solves the problem you care about?

davemgreen committed:
[DAG] Fold nested add(add(reduce(a), b), add(reduce(c), d))

This patch reassociates add(add(vecreduce(a), b), add(vecreduce(c), d)) into
add(vecreduce(add(a, c)), add(b, d)), to combine the reductions into a single
node. This comes up after unrolling vectorized loops.

There is another small change to move reassociateReduction inside fadd outside
of a AllowNewConst block, as new constants will not be created and it should be
OK to perform the combine later after legalization.

david-arm commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -17429,12 +17463,15 @@ SDValue DAGCombiner::visitFADD(SDNode *N) {
+  } // enable-unsafe-fp-math && AllowNewConst
 
+  if (((Options.UnsafeFPMath && Options.NoSignedZerosFPMath) ||
+       (Flags.hasAllowReassociation() && Flags.hasNoSignedZeros()))) {
```
This used to be guarded by `AllowNewConst`, which is defined as

```
  // No FP constant should be created after legalization as Instruction
  // Selection pass has a hard time dealing with FP constants.
  bool AllowNewConst = (Level < AfterLegalizeDAG);
```

With your change we may now potentially create new constants after legalisation. In the new pattern you've added to reassociateReduction it's possible that `B` and `D` are constants and so `op(B, D)` produces a new constant. Is this change required for your optimisation to work? I'm also worried about the existing pattern in `reassociateReduction` which now changes behaviour after removing the `AllowNewConst` guard.

davemgreen replied:
Hi - the last stage of DAG combining, after operations are legalized, will perform legalization on any newly created nodes to make sure nothing illegal is generated into selection. If the fp imm is not legal for the target, but the original two inputs to the are, it still gets a chance to legalize them. Otherwise you would imagine we would see many more cases as immediate can be created from a lot of places (and it would be the fadd(C1, C2) fold that was performing the incorrect fold). I feel doubtful it is possible to have that happen - that all the nodes are not optimized/reassociated earlier but it does constant fold after legalization. Some targets legalize most fp imm's to some from of wrapper / load, so constant folding isn't possible that late.

I think that AllowNewConst is either just a heuristic to prevent worse code, or is old and no longer applies. It is hard to tell as even if I set AllowNewConst=true nothing changes.

david-arm reviewed (approved):
LGTM.
